src/tecrys/svc/world/ContextBaseMusicPlayer.kt

```diff
@@ -3,50 +3,145 @@ package tecrys.svc.world
 import tecrys.svc.world.fleets.FleetSpawner
 
 class ContextBaseMusicPlayer: EveryFrameScript {
+    enum class MusicID(val id: String) {
+        SVC_WHALE_THEME("svc_whale_theme"),
+        SVC_VOIDLING_EXPLORATION_THEME("svc_voidling_exploration_theme"),
+        SVC_VOIDLING_BATTLE_THEME("svc_voidling_battle_theme"),
+        SVC_VOIDLING_BATTLE_THEME_GLITCHED("svc_voidling_battle_theme_glitched");
 
-    private val timer = IntervalUtil(5f, 5f) // check every 5 seconds as to not waste performance
-    private var cooldown = 0f
+        fun toMusicPlayerFormat():String = this.id.lowercase()
+    }
+
+    private data class Music(val json: JSONObject) {
+        private val _music: MutableMap<MusicID, ArrayList<String>> = mutableMapOf()
+
+        init {
+            this.json.keys().forEach { key ->
+                val arr = arrayListOf<String>()
+
+                this.json.getJSONArray(key as String?).forEach { it: JSONObject ->
+                    arr.add(it.getString("file"))
+                }
+
+                _music[MusicID.valueOf((key as String).uppercase())] = arr
+            }
+        }
+
+        fun isMusicPlaying(musicID: MusicID): Boolean {
+            return _music.getOrElse(musicID) { arrayListOf("") }.contains(Global.getSoundPlayer().currentMusicId)
+        }
+
+    }
 
-    companion object{
-        const val COOLDOWN_SECS = 500f // don't play the theme again if it was recently played
+    private val music: Music = try {
+        Music(Global.getSettings().loadJSON("data/config/sounds.json", "symbiotic_void_creatures").getJSONObject("music"));
+    } catch (e: Exception){
+        println(e)
+        throw Exception("Failed to load music files for Symbiotic Void Creatures")
     }
 
+    var lastCheckTime = -10f;
+    var lastCheckState: GameState = GameState.TITLE
+
     override fun isDone(): Boolean = false
 
     override fun runWhilePaused(): Boolean = false
 
     override fun advance(amount: Float) {
-        if(Global.getCurrentState() != GameState.CAMPAIGN) return
-        stopBattleThemeIfApplicable()
-        if(Global.getSector().playerFleet.isInHyperspace) return
-        timer.advance(amount)
-        if(timer.intervalElapsed()){
-            cooldown -= timer.intervalDuration
-            if(cooldown >= 0f) return
-            if(FleetSpawner.getFactionFleets(SVC_FACTION_ID).any {
-                it.containingLocation == Global.getSector().playerFleet.containingLocation
+        if (Global.getCurrentState() == GameState.TITLE) {
+            return
+        }
+
+        val currentTime = Global.getCombatEngine().getTotalElapsedTime(true)
+        val currentState = Global.getCurrentState()
+
+        if (lastCheckState != currentState) {
+            lastCheckTime = currentTime - 10f
+            lastCheckState = currentState
+        }
+
+        when(currentState){
+            GameState.CAMPAIGN -> {
+                stopBattleTheme()
```

Comment:
in here, you don't do a check for lastCheckTime. I.e. in the campaign, you do costly calls (such as e.g. isAnyVoidlingFleetInDistanceHyperspace or FleetSpawner.getFactionFleets) every single frame. That seems like a bad idea for a music player

Reply:
Maybe a LocationChanged-listener would make more sense here?